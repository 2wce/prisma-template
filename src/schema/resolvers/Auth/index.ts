import Mutation from './Mutation'

export default {
  Mutation,
}
