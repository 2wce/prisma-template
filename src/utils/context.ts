import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()
export interface Context {
  prisma: PrismaClient
  request: any
}

export function createContext(request: any) {
  return {
    ...request,
    prisma,
  }
}
