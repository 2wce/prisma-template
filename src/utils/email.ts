import { createTransport } from 'nodemailer'
import { formatError } from './helper'

export const sendEmail = async (
  recipientEmail: string,
  html: string,
  subject: string,
) => {
  try {
    // create reusable transporter object using the default SMTP transport
    const transporter = createTransport({
      host: process.env.SMTP_HOST,
      port: process.env.SMTP_PORT,
      secure: true, // true for 465, false for other ports
      auth: {
        user: process.env.SMTP_USERNAME, // generated ethereal user
        pass: process.env.SMTP_PASSWORD, // generated ethereal password
      },
    })

    // send mail with defined transport object
    const info = await transporter.sendMail({
      from: `Admin <${process.env.NO_REPLY_EMAIL}>`, // sender address
      to: recipientEmail, // list of receivers
      subject, // Subject line
      html, // html body
    })

    console.log('Message sent:', info)

    return info
  } catch (error) {
    formatError('sendEmail', error)
    return error
  }
}
