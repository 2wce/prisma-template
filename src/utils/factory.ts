import { Prisma, PrismaClient } from '@prisma/client'
import * as Factory from 'factory.ts'
import { commerce, internet, name } from 'faker'

const prisma = new PrismaClient()

export interface TableList {
  TABLE_NAME: string
}

export const userFactory = Factory.Sync.makeFactory<Prisma.UserCreateManyInput>(
  {
    name: Factory.Sync.each(() => name.firstName()),
    surname: Factory.Sync.each(() => name.lastName()),
    email: Factory.Sync.each((seq) => `seq_${seq}_${internet.email()}`),
    password: Factory.Sync.each(() => internet.password()),
  },
)

export const postFactory = Factory.Sync.makeFactory<Prisma.PostCreateManyInput>(
  {
    title: Factory.Sync.each(() => commerce.productName()),
    content: Factory.Sync.each(() => commerce.productDescription()),
    published: true,
  },
)

export const clearData = async () => {
  try {
    const posts = await prisma.post.deleteMany()
    const users = await prisma.user.deleteMany()

    console.log({ posts, users })
  } catch (error) {
    console.error(error)
    return error
  }
}
