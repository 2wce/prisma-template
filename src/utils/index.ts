export * from './helper'
export * from './jwt'
export * from './prisma'
