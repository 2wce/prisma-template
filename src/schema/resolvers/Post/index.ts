import { Context } from '../../../utils'

export default {
  Query: {
    feed: (parent, args, ctx: Context) => {
      return ctx.prisma.post.findMany({
        where: { published: true },
      })
    },
    filterPosts: (parent, args, ctx: Context) => {
      return ctx.prisma.post.findMany({
        where: {
          OR: [
            { title: { contains: args.searchString } },
            { content: { contains: args.searchString } },
          ],
        },
      })
    },
    post: (parent, args, ctx: Context) => {
      return ctx.prisma.post.findUnique({
        where: { id: Number(args.where.id) },
      })
    },
  },
  Mutation: {
    createDraft: (parent, args, ctx) => {
      return ctx.prisma.post.create({
        data: {
          title: args.title,
          content: args.content,
          published: false,
          author: {
            connect: { email: args.authorEmail },
          },
        },
      })
    },
    deleteOnePost: (parent, args, ctx: Context) => {
      return ctx.prisma.post.delete({
        where: { id: Number(args.where.id) },
      })
    },
    publish: (parent, args, ctx: Context) => {
      return ctx.prisma.post.update({
        where: { id: Number(args.id) },
        data: { published: true },
      })
    },
  },
  Post: {
    author: (parent, args, ctx: Context) => {
      return ctx.prisma.post
        .findUnique({
          where: { id: parent.id },
        })
        .author()
    },
  },
}
