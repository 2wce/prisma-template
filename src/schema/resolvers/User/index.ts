import { User } from '@prisma/client'
import { Context } from '../../../utils'
import Query from './Query'

export default {
  Query,
  User: {
    posts: (parent: User, _args: unknown, ctx: Context) => {
      return ctx.prisma.user
        .findUnique({
          where: { id: parent.id },
        })
        .posts()
    },
  },
}
