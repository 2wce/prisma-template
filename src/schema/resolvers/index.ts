import { merge } from 'lodash'
import Post from './Post'
import User from './User'

export default merge({}, Post, User)
