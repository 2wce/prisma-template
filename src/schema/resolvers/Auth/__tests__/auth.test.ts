import { Context, createMockContext, MockContext } from '../../../../utils'
import resolvers from '../index'

let mockCtx: MockContext
let context: Context

beforeEach(() => {
  mockCtx = createMockContext()
  context = mockCtx as unknown as Context
})

describe('Auth - Login', () => {
  it('returns an error if invalid input is parsed', async () => {
    // setup
    const {
      Query: { login },
    } = resolvers

    const args = { email: '' }

    // test
    const result = await login({}, args, context)

    // assert
    expect(result).toMatchSnapshot()
  })

  it('returns an error if only email is parsed', async () => {
    // setup
    const {
      Query: { login },
    } = resolvers

    const args = { email: 't@g.com' }

    // test
    const result = await login({}, args, context)

    // assert
    expect(result).toMatchSnapshot()
  })

  it('returns an error if email has invalid email format', async () => {
    // setup
    const {
      Query: { login },
    } = resolvers

    const args = { email: 't' }

    // test
    const result = await login({}, args, context)

    // assert
    expect(result).toMatchSnapshot()
  })
})
