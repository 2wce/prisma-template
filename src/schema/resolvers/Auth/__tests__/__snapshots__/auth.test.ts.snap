// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth - Login returns an error if email has invalid email format 1`] = `[AuthenticationError: Invalid Email Format]`;

exports[`Auth - Login returns an error if invalid input is parsed 1`] = `[AuthenticationError: Please provide your phone number or your e-mail.]`;

exports[`Auth - Login returns an error if only email is parsed 1`] = `[AuthenticationError: User not found.]`;
